(model-revision-notebook)=
# La revisione del modello 

In [1]:
source("_common.R")
suppressPackageStartupMessages({
    library("lavaan")
    library("effectsize")
})

set.seed(42)

{cite:t}`brown2015confirmatory` discute alcune possibili cause che possono essere responsabili della mancanza di adattamento del modello EFA o CFA ai dati. In particolare, vengono esaminate le seguenti possibili cause: 

- il ricercatore ha ipotizzato il numero sbagliato di fattori comuni latenti, 
- un item viene ipotizzato saturare su un solo fattore comune mentre satura su diversi fattori,
- un item viene ipotizzato saturare sul fattore comune sbagliato, 
- è possibile che vi siano correlazioni residue che il modello non ha considerato.

{cite:t}`brown2015confirmatory` mostra come il ricercatore possa usare i *Modification Indices* per valutare le cause del mancato adattamento del modello ai dati.

## Un numero di fattori troppo piccolo

Una delle possibili fonti di mancanza di adattamento del modello può dipendere dal fatto che è stato ipotizzato un numero insufficiente di fattori latenti comuni. {cite:t}`brown2015confirmatory` discute il caso nel quale si confrontano gli indici di bontà di adattamento di un modello ad un solo fattore comune e un modello a due fattori comuni. L'esempio riguarda i dati già in precedenza discussi e relativi relativi a otto misure di personalità raccolte su un campione di 250 pazienti che hanno concluso un programma di psicoterapia. Le scale sono le seguenti:

- anxiety (N1), 
- hostility (N2), 
- depression (N3), 
- self-consciousness (N4), 
- warmth (E1), 
- gregariousness (E2), 
- assertiveness (E3), 
- positive emotions (E4). 

Leggiamo i dati in $\mathsf{R}$.

In [2]:
varnames <- c("N1", "N2", "N3", "N4", "E1", "E2", "E3", "E4")

sds <- c(5.7,  5.6,  6.4,  5.7,  6.0,  6.2,  5.7,  5.6)

cors <- '
 1.000
 0.767  1.000 
 0.731  0.709  1.000 
 0.778  0.738  0.762  1.000 
-0.351  -0.302  -0.356  -0.318  1.000 
-0.316  -0.280  -0.300  -0.267  0.675  1.000 
-0.296  -0.289  -0.297  -0.296  0.634  0.651  1.000 
-0.282  -0.254  -0.292  -0.245  0.534  0.593  0.566  1.000'

psychot_cor_mat <- getCov(cors, names = varnames)

n <- 250

Supponiamo di adattare ai dati il modello "sbagliato" che include un unico fattore comune.  Svolgiamo qui l'analisi *fattoriale esplorativa* usando la funzione sperimentale `efa()` di `lavaan`.

In [3]:
# 1-factor model
f1 <- '
  efa("efa")*f1 =~ N1 + N2 + N3 + N4 + E1 + E2 + E3 + E4
'

 Adattiamo il modello ai dati.

In [4]:
efa_f1 <-
  cfa(
    model = f1,
    sample.cov = psychot_cor_mat,
    sample.nobs = 250,
    rotation = "oblimin"
  )

Consideriamo ora un modello a due fattori.

In [5]:
f2 <- '
  efa("efa")*f1 +
  efa("efa")*f2 =~ N1 + N2 + N3 + N4 + E1 + E2 + E3 + E4
'

Adattiamo il modello ai dati.

In [6]:
efa_f2 <-
  cfa(
    model = f2,
    sample.cov = psychot_cor_mat,
    sample.nobs = 250,
    rotation = "oblimin"
  )

Esaminiamo gli indici di bontà di adattamento.

In [7]:
# define the fit measures
fit_measures_robust <- c("chisq", "df", "pvalue", "cfi", "tli", "rmsea", "srmr")

# collect them for each model
rbind(
  fitmeasures(efa_f1, fit_measures_robust),
  fitmeasures(efa_f2, fit_measures_robust)
) %>%
  # wrangle
  data.frame() %>%
  mutate(
    chisq = round(chisq, digits = 0),
    df = as.integer(df),
    pvalue = ifelse(pvalue == 0, "< .001", pvalue)
  ) %>%
  mutate_at(vars(cfi:srmr), ~ round(., digits = 3))

chisq df pvalue            cfi  tli   rmsea srmr 
1 375   20 < .001            0.71 0.594 0.267 0.187
2  10   13 0.709310449320062 1.00 1.006 0.000 0.010

In [8]:
effectsize::interpret(efa_f1)

Name  Value     Threshold Interpretation
1  GFI   0.6713421 0.95      poor          
2  AGFI  0.4084158 0.90      poor          
3  NFI   0.7006460 0.90      poor          
4  NNFI  0.5941736 0.90      poor          
5  CFI   0.7101240 0.90      poor          
6  RMSEA 0.2665811 0.05      poor          
7  SRMR  0.1873289 0.08      poor          
8  RFI   0.5809044 0.90      poor          
9  PNFI  0.5004614 0.50      satisfactory  
10 IFI   0.7120036 0.90      poor

In [9]:
effectsize::interpret(efa_f2)

Name  Value       Threshold Interpretation
1  GFI   0.990554109 0.95      satisfactory  
2  AGFI  0.973842148 0.90      satisfactory  
3  NFI   0.992174918 0.90      satisfactory  
4  NNFI  1.005603388 0.90      satisfactory  
5  CFI   1.000000000 0.90      satisfactory  
6  RMSEA 0.000000000 0.05      satisfactory  
7  SRMR  0.009907613 0.08      satisfactory  
8  RFI   0.983145977 0.90      satisfactory  
9  PNFI  0.460652640 0.50      poor          
10 IFI   1.002570123 0.90      satisfactory

I risultati mostrano come, in un modello EFA, una soluzione a due fattori produca un adattamento adeguato, mentre ciò non si verifica con un modello ad un solo fattore.

## Specificazione errata delle relazioni tra indicatori e fattori latenti

Un'altra potenziale fonte di errata specificazione del modello CFA è una designazione errata delle relazioni tra indicatori e fattori latenti.

In questo esempio, un ricercatore ha sviluppato un questionario di 12 item (gli item sono valutati su scale da 0 a 8) progettato per valutare le motivazioni dei giovani adulti a consumare bevande alcoliche (Cooper, 1994). La misura aveva lo scopo di valutare tre aspetti di questo costrutto (4 item ciascuno): (1) motivazioni di coping (item 1–4), (2) motivazioni sociali (item 5–8) e (3) motivazioni di miglioramento (item 9 –12). I dati sono i seguenti.

In [10]:
sds <- c(2.06, 1.52, 1.92, 1.41, 1.73, 1.77, 2.49, 2.27, 2.68, 1.75, 2.57, 2.66)

cors <- '
  1.000 
  0.300  1.000 
  0.229  0.261  1.000 
  0.411  0.406  0.429  1.000 
  0.172  0.252  0.218  0.481  1.000 
  0.214  0.268  0.267  0.579  0.484  1.000 
  0.200  0.214  0.241  0.543  0.426  0.492  1.000 
  0.185  0.230  0.185  0.545  0.463  0.548  0.522  1.000 
  0.134  0.146  0.108  0.186  0.122  0.131  0.108  0.151  1.000 
  0.134  0.099  0.061  0.223  0.133  0.188  0.105  0.170  0.448  1.000 
  0.160  0.131  0.158  0.161  0.044  0.124  0.066  0.061  0.370  0.350  1.000 
  0.087  0.088  0.101  0.198  0.077  0.177  0.128  0.112  0.356  0.359  0.507  1.000'

covs <- getCov(cors, sds = sds, names = paste("x", 1:12, sep = ""))

Iniziamo con un modello che ipotizza tre fattori comuni latenti correlati, coerentemente con la motivazione che stava alla base della costruzione dello strumento.

In [11]:
model1 <- '
  copingm  =~ x1 + x2 + x3 + x4
  socialm  =~ x5 + x6 + x7 + x8
  enhancem =~ x9 + x10 + x11 + x12
'

Adattiamo il modello ai dati.

In [12]:
fit1 <- cfa(
  model1, 
  sample.cov = covs, 
  sample.nobs = 500, 
  mimic = "mplus"
)

Warning message in lavaan::lavaan(model = model1, sample.cov = covs, sample.nobs = 500, :
“lavaan WARNING:
    sample.mean= argument is missing, but model contains
    mean/intercept parameters.”


Esaminando le misure di adattamento potremmo concludere che il modello è adeguato.

In [13]:
effectsize::interpret(fit1)

Name  Value      Threshold Interpretation
1  GFI   0.97009178 0.95      satisfactory  
2  AGFI  0.94722078 0.90      satisfactory  
3  NFI   0.94785001 0.90      satisfactory  
4  NNFI  0.97102541 0.90      satisfactory  
5  CFI   0.97761054 0.90      satisfactory  
6  RMSEA 0.03745791 0.05      satisfactory  
7  SRMR  0.03438699 0.08      satisfactory  
8  RFI   0.93251177 0.90      satisfactory  
9  PNFI  0.73242955 0.50      satisfactory  
10 IFI   0.97781875 0.90      satisfactory

Tuttavia, un esame più attento mette in evidenza un comportamento anomalo dell'item `x4` e alcune caratteristiche anomale del modello in generale.

In [14]:
standardizedSolution(fit1)

lhs      op rhs      est.std    se         z         pvalue      
1  copingm  =~ x1       0.43164226 0.03913432 11.029763 0.000000e+00
2  copingm  =~ x2       0.43575459 0.03899692 11.174078 0.000000e+00
3  copingm  =~ x3       0.45123780 0.03847027 11.729519 0.000000e+00
4  copingm  =~ x4       0.95323010 0.02446252 38.966968 0.000000e+00
5  socialm  =~ x5       0.63324280 0.03156182 20.063571 0.000000e+00
6  socialm  =~ x6       0.74779892 0.02546753 29.362831 0.000000e+00
7  socialm  =~ x7       0.68994737 0.02856498 24.153608 0.000000e+00
8  socialm  =~ x8       0.72872974 0.02648066 27.519319 0.000000e+00
9  enhancem =~ x9       0.60199381 0.03863571 15.581280 0.000000e+00
10 enhancem =~ x10      0.59737620 0.03879788 15.397135 0.000000e+00
11 enhancem =~ x11      0.66071892 0.03674260 17.982369 0.000000e+00
12 enhancem =~ x12      0.66516965 0.03661334 18.167412 0.000000e+00
13 x1       ~~ x1       0.81368496 0.03378405 24.084882 0.000000e+00
14 x2       ~~ x2       0.81011794 0.03398617 23.836695 0.000000e+00
15 x3       ~~ x3       0.79638445 0.03471848 22.938343 0.000000e+00
16 x4       ~~ x4       0.09135237 0.04663681  1.958804 5.013577e-02
17 x5       ~~ x5       0.59900356 0.03997259 14.985358 0.000000e+00
18 x6       ~~ x6       0.44079677 0.03808919 11.572753 0.000000e+00
19 x7       ~~ x7       0.52397262 0.03941667 13.293174 0.000000e+00
20 x8       ~~ x8       0.46895296 0.03859449 12.150775 0.000000e+00
21 x9       ~~ x9       0.63760346 0.04651692 13.706916 0.000000e+00
22 x10      ~~ x10      0.64314168 0.04635386 13.874609 0.000000e+00
23 x11      ~~ x11      0.56345051 0.04855306 11.604841 0.000000e+00
24 x12      ~~ x12      0.55754934 0.04870817 11.446732 0.000000e+00
25 copingm  ~~ copingm  1.00000000 0.00000000        NA           NA
26 socialm  ~~ socialm  1.00000000 0.00000000        NA           NA
27 enhancem ~~ enhancem 1.00000000 0.00000000        NA           NA
28 copingm  ~~ socialm  0.79871805 0.03054429 26.149508 0.000000e+00
29 copingm  ~~ enhancem 0.32194358 0.05081203  6.335971 2.358509e-10
30 socialm  ~~ enhancem 0.26764339 0.05555677  4.817476 1.453857e-06
31 x1       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
32 x2       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
33 x3       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
34 x4       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
35 x5       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
36 x6       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
37 x7       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
38 x8       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
39 x9       ~1          0.00000000 0.04472136  0.000000 1.000000e+00
40 x10      ~1          0.00000000 0.04472136  0.000000 1.000000e+00
41 x11      ~1          0.00000000 0.04472136  0.000000 1.000000e+00
42 x12      ~1          0.00000000 0.04472136  0.000000 1.000000e+00
43 copingm  ~1          0.00000000 0.00000000        NA           NA
44 socialm  ~1          0.00000000 0.00000000        NA           NA
45 enhancem ~1          0.00000000 0.00000000        NA           NA
   ci.lower      ci.upper  
1   3.549404e-01 0.50834412
2   3.593220e-01 0.51218714
3   3.758375e-01 0.52663815
4   9.052845e-01 1.00117576
5   5.713828e-01 0.69510283
6   6.978835e-01 0.79771437
7   6.339610e-01 0.74593371
8   6.768286e-01 0.78063089
9   5.262692e-01 0.67771841
10  5.213338e-01 0.67341864
11  5.887048e-01 0.73273309
12  5.934088e-01 0.73693048
13  7.474694e-01 0.87990049
14  7.435063e-01 0.87672961
15  7.283375e-01 0.86443142
16 -5.410864e-05 0.18275884
17  5.206587e-01 0.67734839
18  3.661433e-01 0.51545021
19  4.467174e-01 0.60122788
20  3.933091e-01 0.54459677
21  5.464320e-01 0.72877494
22  5.522898e-01 0.73399357
23  4.682883e-01 0.65861275
24  4.620831e-01 0.65301559
25  1.000000e+00 1.00000000
26  1.000000e+00 1.00000000
27  1.000000e+00 1.00000000
28  7.388524e-01 0.85858375
29  2.223538e-01 

In particolare, l'item `x4` mostra una saturazione molto forte sul fattore Motivi di coping (.955) ed emerge una correlazione molto alta tra i fattori Motivi di coping e Motivi sociali (.798).

@brown2015confirmatory suggerisce di esaminare i *Modification Indices*. Tale esame mostra che il MI associato a `x4` è molto alto, 18.916.

In [15]:
modindices(fit1)

lhs      op rhs mi           epc          sepc.lv      sepc.all    
46  copingm  =~ x5   0.030005989 -0.029864968 -0.026528841 -0.015349946
47  copingm  =~ x6   0.483607013  0.126715141  0.112560167  0.063657017
48  copingm  =~ x7   0.779713184  0.220132468  0.195542120  0.078609654
49  copingm  =~ x8   1.962276016 -0.323442154 -0.287311386 -0.126695694
50  copingm  =~ x9   0.101119531  0.044105439  0.039178551  0.014633498
51  copingm  =~ x10  2.016179576  0.128686040  0.114310902  0.065385941
52  copingm  =~ x11  1.870017679 -0.181368830 -0.161108654 -0.062750981
53  copingm  =~ x12  0.039774895 -0.027386556 -0.024327285 -0.009154759
54  socialm  =~ x1   6.926785939 -0.520304949 -0.569429069 -0.276698865
55  socialm  =~ x2   0.052053171 -0.033339419 -0.036487130 -0.024028734
56  socialm  =~ x3   2.058455483 -0.266795427 -0.291984676 -0.152227711
57  socialm  =~ x4  18.916304911  1.300066952  1.422811595  1.010097106
58  socialm  =~ x9   0.338261984  0.066907544  0.073224559  0.027349952
59  socialm  =~ x10  2.883617084  0.127660732  0.139713704  0.079916367
60  socialm  =~ x11  4.357439140 -0.229221561 -0.250863307 -0.097709951
61  socialm  =~ x12  0.001070604  0.003719464  0.004070634  0.001531847
62  enhancem =~ x1   1.954056978  0.092553029  0.149170471  0.072485411
63  enhancem =~ x2   0.862633625  0.045296430  0.073005605  0.048078110
64  enhancem =~ x3   0.379815944  0.037713503  0.060783976  0.031690038
65  enhancem =~ x4   3.101582135 -0.103926028 -0.167500673 -0.118913808
66  enhancem =~ x5   0.595829582 -0.038973519 -0.062814781 -0.036345482
67  enhancem =~ x6   2.495315594  0.077609406  0.125085390  0.070740502
68  enhancem =~ x7   0.538673545 -0.051861381 -0.083586531 -0.033602521
69  enhancem =~ x8   0.093095151 -0.019336415 -0.031165076 -0.013742863
70  x1       ~~ x2  10.298966995  0.379140628  0.379140628  0.149436604
71  x1       ~~ x3   0.985767890  0.147339714  0.147339714  0.046369423
72  x1       ~~ x4   0.015598853 -0.014806108 -0.014806108 -0.018734251
73  x1       ~~ x5   0.451852422 -0.080302627 -0.080302627 -0.032340240
74  x1       ~~ x6   0.484485902 -0.077883982 -0.077883982 -0.035737988
75  x1       ~~ x7   0.289527960 -0.088717149 -0.088717149 -0.026541682
⋮   ⋮        ⋮  ⋮   ⋮            ⋮            ⋮            ⋮           
106 x4       ~~ x11  2.637247586 -0.14863761  -0.14863761  -0.181158325
107 x4       ~~ x12  0.169101191  0.03892761   0.03892761   0.046081350
108 x5       ~~ x6   0.369571094  0.05700901   0.05700901   0.036304439
109 x5       ~~ x7   0.292001629 -0.07226591  -0.07226591  -0.030004653
110 x5       ~~ x8   0.007087368  0.01015175   0.01015175   0.004887190
111 x5       ~~ x9   0.822203615  0.13325873   0.13325873   0.046600871
112 x5       ~~ x10  0.339106523  0.05600380   0.05600380   0.029863092
113 x5       ~~ x11  1.125776934 -0.14545894  -0.14545894  -0.056427145
114 x5       ~~ x12  1.143486321 -0.15142439  -0.15142439  -0.057053374
115 x6       ~~ x7   2.528078433 -0.21453919  -0.21453919  -0.101491633
116 x6       ~~ x8   0.052656710  0.02855951   0.02855951   0.015665253
117 x6       ~~ x9   1.055994918 -0.14050502  -0.14050502  -0.055983319
118 x6       ~~ x10  0.598393109  0.06920893   0.06920893   0.042048247
119 x6       ~~ x11  0.248120056  0.06361858   0.06361858   0.028119014
120 x6       ~~ x12  1.667231850  0.17036266   0.17036266   0.073135524
121 x7       ~~ x8   1.430701442  0.20641391   0.20641391   0.073818398
122 x7       ~~ x9   0.031974082 -0.03617847  -0.03617847  -0.009398444
123 x7       ~~ x10  1.520595023 -0.16326316  -0.16326316  -0.064671491
124 x7       ~~ x11  0.262702017 -0.09678497  -0.09678497  -0.027890928
125 x7       ~~ x12  0.637279724  0.15571434   0.15571434   0.043583441
126 x8       ~~ x9   1.621415958  0.22713544   0.22713544   0.068415381
127 x8       ~~ x10  1.311207860  0.13365654   0.13365654   0.061387222
128 x8       ~~ x11  2.143803609 -0.24388059  -0.24388059  -0.081488322
129 x8       ~~ x12  0.590763954 -0.13225157

Le considerazioni precedenti, dunque, suggeriscono che il modello potrebbe non avere descritto in maniera adeguata le relazioni tra `x4` e i fattori comuni latenti.  In base a considerazioni teoriche, supponiamo che abbia senso pensare che `x4` saturi non solo sul fattore Motivi di coping ma anche sul fattore di Motivi Sociali. Specifichiamo dunque un nuovo modello nel modo seguente.

In [16]:
model2 <- '
  copingm  =~ x1 + x2 + x3 + x4
  socialm  =~ x4 + x5 + x6 + x7 + x8
  enhancem =~ x9 + x10 + x11 + x12
'

Adattiamo il modello.

In [17]:
fit2 <- cfa(
  model2, 
  sample.cov = covs, 
  sample.nobs = 500, 
  mimic = "mplus"
)

Warning message in lavaan::lavaan(model = model2, sample.cov = covs, sample.nobs = 500, :
“lavaan WARNING:
    sample.mean= argument is missing, but model contains
    mean/intercept parameters.”


Esaminiamo gli indici di bontà di adattamento.

In [18]:
effectsize::interpret(fit2)

Name  Value      Threshold Interpretation
1  GFI   0.97684139 0.95      satisfactory  
2  AGFI  0.95831451 0.90      satisfactory  
3  NFI   0.95826773 0.90      satisfactory  
4  NNFI  0.98393923 0.90      satisfactory  
5  CFI   0.98783275 0.90      satisfactory  
6  RMSEA 0.02788804 0.05      satisfactory  
7  SRMR  0.02887855 0.08      satisfactory  
8  RFI   0.94491340 0.90      satisfactory  
9  PNFI  0.72596040 0.50      satisfactory  
10 IFI   0.98795337 0.90      satisfactory

La bontà di adattamento è migliorata.

Esaminiamo la soluzione standardizzata. Vediamo ora che sono scomparse le due anomalie trovate in precedenza.

In [19]:
standardizedSolution(fit2)

lhs      op rhs      est.std   se         z         pvalue       ci.lower   
1  copingm  =~ x1       0.5136919 0.04268561 12.034311 0.000000e+00  0.43002967
2  copingm  =~ x2       0.5149158 0.04265451 12.071778 0.000000e+00  0.43131452
3  copingm  =~ x3       0.5160398 0.04262607 12.106202 0.000000e+00  0.43249424
4  copingm  =~ x4       0.5380264 0.06212680  8.660133 0.000000e+00  0.41626007
5  socialm  =~ x4       0.4389970 0.06093489  7.204362 5.830891e-13  0.31956680
6  socialm  =~ x5       0.6318747 0.03160090 19.995465 0.000000e+00  0.56993804
7  socialm  =~ x6       0.7464621 0.02549441 29.279444 0.000000e+00  0.69649398
8  socialm  =~ x7       0.6905804 0.02849475 24.235357 0.000000e+00  0.63473172
9  socialm  =~ x8       0.7308592 0.02632560 27.762304 0.000000e+00  0.67926197
10 enhancem =~ x9       0.6026163 0.03856640 15.625422 0.000000e+00  0.52702753
11 enhancem =~ x10      0.5946609 0.03884764 15.307516 0.000000e+00  0.51852093
12 enhancem =~ x11      0.6649364 0.03655910 18.187989 0.000000e+00  0.59328193
13 enhancem =~ x12      0.6629023 0.03661848 18.102943 0.000000e+00  0.59113140
14 x1       ~~ x1       0.7361206 0.04385451 16.785517 0.000000e+00  0.65016735
15 x2       ~~ x2       0.7348617 0.04392697 16.729169 0.000000e+00  0.64876641
16 x3       ~~ x3       0.7337029 0.04399350 16.677531 0.000000e+00  0.64747725
17 x4       ~~ x4       0.2298436 0.03653228  6.291520 3.143723e-10  0.15824160
18 x5       ~~ x5       0.6007344 0.03993562 15.042573 0.000000e+00  0.52246204
19 x6       ~~ x6       0.4427943 0.03806122 11.633740 0.000000e+00  0.36819571
20 x7       ~~ x7       0.5230987 0.03935583 13.291518 0.000000e+00  0.44596271
21 x8       ~~ x8       0.4658448 0.03848061 12.105964 0.000000e+00  0.39042424
22 x9       ~~ x9       0.6368536 0.04648148 13.701233 0.000000e+00  0.54575157
23 x10      ~~ x10      0.6463784 0.04620235 13.990164 0.000000e+00  0.55582348
24 x11      ~~ x11      0.5578595 0.04861895 11.474117 0.000000e+00  0.46256814
25 x12      ~~ x12      0.5605605 0.04854895 11.546295 0.000000e+00  0.46540634
26 copingm  ~~ copingm  1.0000000 0.00000000        NA           NA  1.00000000
27 socialm  ~~ socialm  1.0000000 0.00000000        NA           NA  1.00000000
28 enhancem ~~ enhancem 1.0000000 0.00000000        NA           NA  1.00000000
29 copingm  ~~ socialm  0.6096010 0.05673806 10.744128 0.000000e+00  0.49839643
30 copingm  ~~ enhancem 0.3499566 0.05867859  5.963957 2.462012e-09  0.23494868
31 socialm  ~~ enhancem 0.2645008 0.05517209  4.794105 1.634027e-06  0.15636549
32 x1       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
33 x2       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
34 x3       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
35 x4       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
36 x5       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
37 x6       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
38 x7       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
39 x8       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
40 x9       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
41 x10      ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
42 x11      ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
43 x12      ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
44 copingm  ~1          0.0000000 0.00000000        NA           NA  0.00000000
45 socialm  ~1          0.0000000 0.00000000        NA           NA  0.00000000
46 enhancem ~1          0.0000000 0.00000000        NA           NA  0.00000000
   ci.upper  
1  0.59735419
2  0.59851714
3  0.59958536
4  0.65979267
5  0.55842720
6  0.69381129
7  0.79643023
8  0.74642908
9  0.78245641
10 0.67820505
11 0.67080088
12 0.73659095
13 0.73467321
14 0.82207386
15 0.82095697
16 0.81992859
17 0.

Esaminando i MI, notiamo che il modello potrebbe migliorare se introduciamo una correlazione tra le specificità `x11` e `x12`.

In [20]:
modindices(fit2)

lhs      op rhs mi           epc          sepc.lv      sepc.all    
47  copingm  =~ x5  0.075879281   0.032464745  0.034320003  0.019858025
48  copingm  =~ x6  1.412911873   0.143137105  0.151316941  0.085575383
49  copingm  =~ x7  0.244623644   0.083197038  0.087951488  0.035357264
50  copingm  =~ x8  3.668026217  -0.294652924 -0.311491413 -0.137358226
51  copingm  =~ x9  0.242920085   0.065568420  0.069315449  0.025889887
52  copingm  =~ x10 0.566100546   0.065434430  0.069173802  0.039567476
53  copingm  =~ x11 0.119109121  -0.043934458 -0.046445174 -0.018090153
54  copingm  =~ x12 0.598292941  -0.101897430 -0.107720548 -0.040537007
55  socialm  =~ x1  1.947767101  -0.395605353 -0.244629140 -0.118870915
56  socialm  =~ x2  0.718049453   0.177418638  0.109709761  0.072249732
57  socialm  =~ x3  0.297752655   0.144452800  0.089324788  0.046569913
58  socialm  =~ x9  0.315538712   0.114063271  0.070532918  0.026344621
59  socialm  =~ x10 3.169241286   0.236374679  0.146166208  0.083607200
60  socialm  =~ x11 4.927007444  -0.430218894 -0.266032998 -0.103618466
61  socialm  =~ x12 0.017124127   0.026249184  0.016231619  0.006108224
62  enhancem =~ x1  0.313590008   0.039551624  0.063812400  0.031007910
63  enhancem =~ x2  0.002792433   0.002753742  0.004442873  0.002925869
64  enhancem =~ x3  0.036819178  -0.012629957 -0.020377113 -0.010623707
65  enhancem =~ x4  0.106389675  -0.012820452 -0.020684455 -0.014684515
66  enhancem =~ x5  0.464252320  -0.034063087 -0.054957221 -0.031799002
67  enhancem =~ x6  2.703472740   0.079190634  0.127765789  0.072256327
68  enhancem =~ x7  0.466547100  -0.047570946 -0.076750737 -0.030854464
69  enhancem =~ x8  0.094615783  -0.019129220 -0.030862991 -0.013609639
70  x1       ~~ x2  1.966143329   0.187287831  0.187287831  0.081487437
71  x1       ~~ x3  2.042378382  -0.241327807 -0.241327807 -0.083190475
72  x1       ~~ x4  0.775485559   0.098280683  0.098280683  0.082425285
73  x1       ~~ x5  0.237902661  -0.057632489 -0.057632489 -0.024367321
74  x1       ~~ x6  0.186730495  -0.047855957 -0.047855957 -0.023035053
75  x1       ~~ x7  0.018723943  -0.022287044 -0.022287044 -0.007015991
76  x1       ~~ x8  0.365508385  -0.086948141 -0.086948141 -0.031815678
⋮   ⋮        ⋮  ⋮   ⋮            ⋮            ⋮            ⋮           
106 x4       ~~ x11  1.312840720 -0.101878937 -0.101878937 -0.078672506
107 x4       ~~ x12  0.321917677  0.052251350  0.052251350  0.038890072
108 x5       ~~ x6   0.503619921  0.066410080  0.066410080  0.042134879
109 x5       ~~ x7   0.262162750 -0.068368943 -0.068368943 -0.028369388
110 x5       ~~ x8   0.004247385  0.007840557  0.007840557  0.003781659
111 x5       ~~ x9   0.849725569  0.135425450  0.135425450  0.047318183
112 x5       ~~ x10  0.288139536  0.051689288  0.051689288  0.027453727
113 x5       ~~ x11  1.018598997 -0.138036785 -0.138036785 -0.053738004
114 x5       ~~ x12  1.223895837 -0.156755686 -0.156755686 -0.058818324
115 x6       ~~ x7   2.404267265 -0.208536197 -0.208536197 -0.098511143
116 x6       ~~ x8   0.033517700  0.022710567  0.022710567  0.012470280
117 x6       ~~ x9   0.978424122 -0.135107947 -0.135107947 -0.053742946
118 x6       ~~ x10  0.523836252  0.064789637  0.064789637  0.039175910
119 x6       ~~ x11  0.340744066  0.074312372  0.074312372  0.032935224
120 x6       ~~ x12  1.519932821  0.162590927  0.162590927  0.069454190
121 x7       ~~ x8   1.170700334  0.186223650  0.186223650  0.066875403
122 x7       ~~ x9   0.019578625 -0.028247666 -0.028247666 -0.007348632
123 x7       ~~ x10  1.593217056 -0.167012340 -0.167012340 -0.066045841
124 x7       ~~ x11  0.175352432 -0.078735934 -0.078735934 -0.022822112
125 x7       ~~ x12  0.585734957  0.149079342  0.149079342  0.041648846
126 x8       ~~ x9   1.807601677  0.238879668  0.238879668  0.072234959
127 x8       ~~ x10  1.267267918  0.131087814  0.131087814  0.060256451
128 x8       ~~ x11  1.790880844 -0.221558423 -0.221558423 -0.074647528
129 x8       ~~ x12  0.594705739 -0.13226575

Il nuovo modello diventa dunque il seguente.

In [21]:
model3 <- '
  copingm  =~ x1 + x2 + x3 + x4
  socialm  =~ x4 + x5 + x6 + x7 + x8
  enhancem =~ x9 + x10 + x11 + x12
  x11 ~~ x12
'

Adattiamo il modello.

In [22]:
fit3 <- cfa(
  model3, 
  sample.cov = covs, 
  sample.nobs = 500, 
  mimic = "mplus"
)

Warning message in lavaan::lavaan(model = model3, sample.cov = covs, sample.nobs = 500, :
“lavaan WARNING:
    sample.mean= argument is missing, but model contains
    mean/intercept parameters.”


Un test basato sul rapporto di verosimiglianze conferma che il miglioramento di adattamento è sostanziale.

In [23]:
lavTestLRT(fit2, fit3)

Df AIC      BIC      Chisq    Chisq diff RMSEA     Df diff Pr(>Chisq)  
fit3 49 23934.34 24107.14 44.95535       NA          NA NA                NA
fit2 50 23956.83 24125.41 69.44357 24.48823   0.2167405  1      7.476528e-07

Esaminiamo gli indici di bontà di adattamento.

In [24]:
out = summary(fit3, fit.measures = TRUE)
print(out)

lavaan 0.6.15 ended normally after 61 iterations

  Estimator                                         ML
  Optimization method                           NLMINB
  Number of model parameters                        41

  Number of observations                           500

Model Test User Model:
                                                      
  Test statistic                                44.955
  Degrees of freedom                                49
  P-value (Chi-square)                           0.638

Model Test Baseline Model:

  Test statistic                              1664.026
  Degrees of freedom                                66
  P-value                                        0.000

User Model versus Baseline Model:

  Comparative Fit Index (CFI)                    1.000
  Tucker-Lewis Index (TLI)                       1.003

Loglikelihood and Information Criteria:

  Loglikelihood user model (H0)             -11926.170
  Loglikelihood unrestricted model (H1)     -119

Gli indici di fit sono migliorati.

Esaminiamo la soluzione standardizzata.

In [25]:
standardizedSolution(fit3)

lhs      op rhs      est.std   se         z         pvalue       ci.lower   
1  copingm  =~ x1       0.5137373 0.04275270 12.016486 0.000000e+00  0.42994349
2  copingm  =~ x2       0.5150194 0.04272080 12.055473 0.000000e+00  0.43128819
3  copingm  =~ x3       0.5144101 0.04273594 12.036943 0.000000e+00  0.43064919
4  copingm  =~ x4       0.5396994 0.06268938  8.609105 0.000000e+00  0.41683052
5  socialm  =~ x4       0.4377501 0.06140696  7.128672 1.013412e-12  0.31739467
6  socialm  =~ x5       0.6319500 0.03159184 20.003584 0.000000e+00  0.57003112
7  socialm  =~ x6       0.7464930 0.02548552 29.290865 0.000000e+00  0.69654231
8  socialm  =~ x7       0.6901567 0.02851159 24.206176 0.000000e+00  0.63427496
9  socialm  =~ x8       0.7311852 0.02630130 27.800344 0.000000e+00  0.67963558
10 enhancem =~ x9       0.6691959 0.04083337 16.388454 0.000000e+00  0.58916395
11 enhancem =~ x10      0.6637281 0.04086182 16.243234 0.000000e+00  0.58364045
12 enhancem =~ x11      0.5418274 0.04470362 12.120436 0.000000e+00  0.45420990
13 enhancem =~ x12      0.5407209 0.04474879 12.083474 0.000000e+00  0.45301484
14 x11      ~~ x12      0.3027101 0.04964764  6.097169 1.079635e-09  0.20540247
15 x1       ~~ x1       0.7360740 0.04392731 16.756638 0.000000e+00  0.64997809
16 x2       ~~ x2       0.7347550 0.04400408 16.697428 0.000000e+00  0.64850858
17 x3       ~~ x3       0.7353823 0.04396760 16.725549 0.000000e+00  0.64920734
18 x4       ~~ x4       0.2288915 0.03678071  6.223138 4.873071e-10  0.15680259
19 x5       ~~ x5       0.6006392 0.03992892 15.042710 0.000000e+00  0.52237997
20 x6       ~~ x6       0.4427482 0.03804953 11.636101 0.000000e+00  0.36817247
21 x7       ~~ x7       0.5236838 0.03935493 13.306687 0.000000e+00  0.44654953
22 x8       ~~ x8       0.4653682 0.03846224 12.099353 0.000000e+00  0.38998362
23 x9       ~~ x9       0.5521769 0.05465105 10.103682 0.000000e+00  0.44506275
24 x10      ~~ x10      0.5594649 0.05424229 10.314185 0.000000e+00  0.45315202
25 x11      ~~ x11      0.7064231 0.04844330 14.582474 0.000000e+00  0.61147596
26 x12      ~~ x12      0.7076210 0.04839321 14.622320 0.000000e+00  0.61277201
27 copingm  ~~ copingm  1.0000000 0.00000000        NA           NA  1.00000000
28 socialm  ~~ socialm  1.0000000 0.00000000        NA           NA  1.00000000
29 enhancem ~~ enhancem 1.0000000 0.00000000        NA           NA  1.00000000
30 copingm  ~~ socialm  0.6099548 0.05682136 10.734604 0.000000e+00  0.49858697
31 copingm  ~~ enhancem 0.3532465 0.06044471  5.844126 5.092334e-09  0.23477709
32 socialm  ~~ enhancem 0.2892191 0.05625324  5.141376 2.727332e-07  0.17896475
33 x1       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
34 x2       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
35 x3       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
36 x4       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
37 x5       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
38 x6       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
39 x7       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
40 x8       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
41 x9       ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
42 x10      ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
43 x11      ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
44 x12      ~1          0.0000000 0.04472136  0.000000 1.000000e+00 -0.08765225
45 copingm  ~1          0.0000000 0.00000000        NA           NA  0.00000000
46 socialm  ~1          0.0000000 0.00000000        NA           NA  0.00000000
47 enhancem ~1          0.0000000 0.00000000        NA           NA  0.00000000
   ci.upper  
1  0.59753101
2  0.59875065
3  0.59817100
4  0.66256838
5  0.55810553
6  0.69386884
7  0.79644372
8  0.74603836
9  0.78273478
10 0.74922784
11 0.7438

Non ci sono ulteriori motivi di preoccupazione.  {cite:t}`brown2015confirmatory` conclude che il modello più adeguato sia `model3`. 

Nel caso presente, a mio parare, l'introduzione della correlazione residua tra `x11` e `x12` si sarebbe anche potuta evitare, dato che il modello `model3` (con meno idiosincrasie legate al campione) si era già dimostrato adeguato.

## Saturazione sul fattore sbagliato

{cite:t}`brown2015confirmatory` considera anche il caso opposto, ovvero quello nel quale il ricercatore ipotizza una saturazione spuria. Per i dati in discussione, si può avere la situazione presente.

In [26]:
model4 <- '
  copingm  =~ x1 + x2 + x3 + x4
  socialm  =~ x4 +x5 + x6 + x7 + x8 + x12
  enhancem =~ x9 + x10 + x11
'

Adattiamo il modello ai dati.

In [27]:
fit4 <- cfa(
  model4, 
  sample.cov = covs, 
  sample.nobs = 500, 
  mimic = "mplus"
)

Warning message in lavaan::lavaan(model = model4, sample.cov = covs, sample.nobs = 500, :
“lavaan WARNING:
    sample.mean= argument is missing, but model contains
    mean/intercept parameters.”


Esaminiamo la soluzione ottenuta.

In [28]:
out = summary(fit4, fit.measures = TRUE)
print(out)

lavaan 0.6.15 ended normally after 59 iterations

  Estimator                                         ML
  Optimization method                           NLMINB
  Number of model parameters                        40

  Number of observations                           500

Model Test User Model:
                                                      
  Test statistic                               212.717
  Degrees of freedom                                50
  P-value (Chi-square)                           0.000

Model Test Baseline Model:

  Test statistic                              1664.026
  Degrees of freedom                                66
  P-value                                        0.000

User Model versus Baseline Model:

  Comparative Fit Index (CFI)                    0.898
  Tucker-Lewis Index (TLI)                       0.866

Loglikelihood and Information Criteria:

  Loglikelihood user model (H0)             -12010.051
  Loglikelihood unrestricted model (H1)     -119

È chiaro che il modello `model4` è inadeguato. Il problema emerge chiaramente anche esaminando i MI.

In [29]:
modindices(fit4)

lhs      op rhs mi           epc          sepc.lv      sepc.all    
47  copingm  =~ x5    0.09017503  0.03562312   0.03784791   0.021899323
48  copingm  =~ x6    0.55421214  0.09003505   0.09565805   0.054098226
49  copingm  =~ x7    0.10746410  0.05543532   0.05889745   0.023677281
50  copingm  =~ x8    3.91880191 -0.30573892  -0.32483339  -0.143241711
51  copingm  =~ x12   6.10941671  0.49879335   0.52994476   0.199427006
52  copingm  =~ x9    0.38979797 -0.09562061  -0.10159245  -0.037945578
53  copingm  =~ x10   0.02663662 -0.01608649  -0.01709114  -0.009776150
54  copingm  =~ x11   0.82324384  0.12309854   0.13078648   0.050940643
55  socialm  =~ x1    1.99011638 -0.39751600  -0.25053040  -0.121738462
56  socialm  =~ x2    0.63818689  0.16639198   0.10486685   0.069060382
57  socialm  =~ x3    0.37217899  0.16023194   0.10098454   0.052648813
58  socialm  =~ x9    0.31477766 -0.13045181  -0.08221592  -0.030708291
59  socialm  =~ x10   1.42307040  0.17864053   0.11258637   0.064399504
60  socialm  =~ x11   0.52007742 -0.14971313  -0.09435517  -0.036750842
61  enhancem =~ x1    1.02911333  0.06669409   0.12072607   0.058663564
62  enhancem =~ x2    0.23155704  0.02334072   0.04225013   0.027823949
63  enhancem =~ x3    0.15333314 -0.02399418  -0.04343298  -0.022644009
64  enhancem =~ x4    0.74519462 -0.03094824  -0.05602084  -0.039770883
65  enhancem =~ x5    0.34312399 -0.02781319  -0.05034594  -0.029130859
66  enhancem =~ x6    0.10312748  0.01464260   0.02650526   0.014989720
67  enhancem =~ x7    2.75182606 -0.10964223  -0.19846852  -0.079786055
68  enhancem =~ x8    0.12939752 -0.02122648  -0.03842304  -0.016943398
69  enhancem =~ x12 116.78055563  0.91578890   1.65771216   0.623824591
70  x1       ~~ x2    1.70856301  0.17666232   0.17666232   0.077186338
71  x1       ~~ x3    2.27308679 -0.25695039  -0.25695039  -0.088754530
72  x1       ~~ x4    0.85019572  0.10317774   0.10317774   0.086051193
73  x1       ~~ x5    0.29156144 -0.06388157  -0.06388157  -0.026974801
74  x1       ~~ x6    0.18845155 -0.04795089  -0.04795089  -0.023176645
75  x1       ~~ x7    0.02307378 -0.02475596  -0.02475596  -0.007785225
76  x1       ~~ x8    0.41852446 -0.09314066  -0.09314066  -0.033984772
⋮   ⋮        ⋮  ⋮   ⋮            ⋮            ⋮            ⋮           
106 x4       ~~ x10 9.138852e-01  0.060508783  0.060508783  0.067730785
107 x4       ~~ x11 5.169850e-01 -0.068349097 -0.068349097 -0.046530997
108 x5       ~~ x6  6.113739e-01  0.072842975  0.072842975  0.046181582
109 x5       ~~ x7  1.153198e-01 -0.045275647 -0.045275647 -0.018675970
110 x5       ~~ x8  7.918745e-02  0.033755738  0.033755738  0.016155480
111 x5       ~~ x12 3.264796e+00 -0.302396718 -0.302396718 -0.086687525
112 x5       ~~ x9  2.028339e-01  0.066152369  0.066152369  0.024957253
113 x5       ~~ x10 4.522542e-04  0.002047507  0.002047507  0.001160392
114 x5       ~~ x11 2.311883e+00 -0.223779997 -0.223779997 -0.077133427
115 x6       ~~ x7  2.238702e+00 -0.199955488 -0.199955488 -0.094411109
116 x6       ~~ x8  7.348780e-02  0.033339051  0.033339051  0.018264036
117 x6       ~~ x12 4.778443e-01  0.108510095  0.108510095  0.035605815
118 x6       ~~ x9  1.251023e+00 -0.152816754 -0.152816754 -0.065992352
119 x6       ~~ x10 7.839316e-01  0.079234586  0.079234586  0.051400286
120 x6       ~~ x11 3.703228e-01  0.082972291  0.082972291  0.032736007
121 x7       ~~ x8  1.644356e+00  0.219424488  0.219424488  0.078210665
122 x7       ~~ x12 4.328632e-01 -0.151939667 -0.151939667 -0.032438394
123 x7       ~~ x9  5.176871e-03 -0.014538491 -0.014538491 -0.004084880
124 x7       ~~ x10 1.836280e+00 -0.179429687 -0.179429687 -0.075732522
125 x7       ~~ x11 3.478588e-01 -0.119232436 -0.119232436 -0.030607258
126 x8       ~~ x12 2.680163e+00 -0.334567892 -0.334567892 -0.082875501
127 x8       ~~ x9  6.757446e-01  0.146530430  0.146530430  0.047768497
128 x8       ~~ x10 3.373531e-01  0.067826260  0.067826260  0.033215418
129 x8       ~~ x11 3.436823e+00 -0.33011454

Il MI relativo alla saturazione di `x12` su `enhancem` è uguale a 116.781. Chiaramente, in una revisione del modello, questo problema dovrebbe deve essere affrontato.

## Commenti e considerazioni finali {-}

Gli esempi discussi da {cite:t}`brown2015confirmatory` mostrano come l'uso dei MI, insieme all'esame della soluzione fattoriale, possano essere usati dallo psicologo per migliorare il modello che viene proposto.
